### PR TITLE
[MAcOS]Fix bugs related to height properties

### PR DESCRIPTION
**Pull request by mukuagar**
## Description

* Fix issue where elements would not get proper spacing when they were initially hidden and then toggled visible
* Fix issue where container would take spacing even when it was completely empty and had no elements
* Fix issue where input label and error messages were hidden but were still increasing height of card

## Sample Card
Case 1
`{
    "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
    "actions": [
        {
            "title": "OK",
            "type": "Action.Submit"
        }
    ],
    "body": [
        {
            "errorMessage": "error message error message error message error message ",
            "height": "stretch",
            "id": "1",
            "isRequired": true,
            "label": "Hello this is label",
            "type": "Input.Date",
            "isVisible": false
        },
        {
            "errorMessage": "error message error message error message error message ",
            "height": "stretch",
            "id": "3",
            "isRequired": true,
            "label": "This is label",
            "type": "Input.Text",
            "isVisible": false
        },
        {
            "actions": [
                {
                    "targetElements": [
                        "1",
                        "3"
                    ],
                    "title": "Action.ToggleVisibility",
                    "type": "Action.ToggleVisibility"
                }
            ],
            "id": "2",
            "type": "ActionSet"
        }
    ],
    "minHeight": "400px",
    "type": "AdaptiveCard",
    "version": "1.3"
}`

Case 2:
`{
    "type": "AdaptiveCard",
    "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
    "version": "1.3",
    "minHeight": "300px",
    "body": [
        {
            "type": "Container",
            "style": "accent",
            "minHeight": "200px",
            "items": [
                {
                    "type": "ColumnSet",
                    "columns": [
                        {
                            "type": "Column",
                            "width": "stretch",
                            "style": "emphasis"
                        },
                        {
                            "type": "Column",
                            "width": "stretch",
                            "style": "good"
                        }
                    ]
                },
                {
                    "type": "Container",
                    "style": "attention"
                },
                {
                    "type": "TextBlock",
                    "wrap": true,
                    "text": "Test Text box"
                }
            ]
        },
        {
            "type": "Container"
        },
        {
            "type": "TextBlock",
            "wrap": true,
            "text": "Test Text box"
        }
    ]
}
`

Case 3:
`{
    "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
    "actions": [
        {
            "title": "OK",
            "type": "Action.Submit"
        }
    ],
    "body": [
        {
            "type": "Container",
            "items": [
                {
                    "errorMessage": "error message error message error message error message ",
                    "height": "stretch",
                    "id": "1",
                    "isRequired": true,
                    "label": "Hello this is label",
                    "type": "Input.Date",
                    "isVisible": false
                },
                {
                    "errorMessage": "error message error message error message error message ",
                    "height": "stretch",
                    "id": "3",
                    "isRequired": true,
                    "label": "This is label",
                    "type": "Input.Time",
                    "isVisible": false
                }
            ]
        },
        {
            "actions": [
                {
                    "targetElements": [
                        "1",
                        "3"
                    ],
                    "title": "Action.ToggleVisibility",
                    "type": "Action.ToggleVisibility"
                }
            ],
            "id": "2",
            "type": "ActionSet"
        }
    ],
    "minHeight": "400px",
    "type": "AdaptiveCard",
    "version": "1.3"
}
`

| old | new |
|----|----|
| uneven-height | uneven-height-fixed |
| Screenshot 2022-11-28 at 1 56 13 PM | Screenshot 2022-11-28 at 1 23 54 PM |

## How Verified
How you verified the fix, including one or all of the following: 
1. Ran BTP for height changes and BTP for toggle visibility

### CODE REVIEW

File: source/macos/AdaptiveCards/AdaptiveCards/Views/ACRColumnSetView.swift [hunk 1]
```diff
@@ -68,8 +68,8 @@ class ACRColumnSetView: ACRContentStackView {
         }
     }
     
-    override func configureLayoutAndVisibility(_ verticalContentAlignment: ACSVerticalContentAlignment, minHeight: NSNumber?, heightType: ACSHeightType, type: ACSCardElementType, isRootMinHeightAvailable: Bool) {
+    override func configureLayoutAndVisibility(element: NSObject, isRootMinHeightAvailable: Bool) {
```
Comment (urmitchauhan) [inline source/macos/AdaptiveCards/AdaptiveCards/Views/ACRColumnSetView.swift:71]:
```suggestion
    override func configureLayoutAndVisibility(element: ACSBaseCardElement, isRootMinHeightAvailable: Bool) {
```

File: source/macos/AdaptiveCards/AdaptiveCards/Views/ACRColumnSetView.swift [hunk 2]
```diff
@@ -68,8 +68,8 @@ class ACRColumnSetView: ACRContentStackView {
...
         self.applyVisibilityToSubviews()
-        self.setMinimumHeight(minHeight)
+        self.setMinimumHeight((element as? ACSCollectionTypeElement)?.getMinHeight())
```
Comment (urmitchauhan) [inline source/macos/AdaptiveCards/AdaptiveCards/Views/ACRColumnSetView.swift:73]:
```suggestion
        self.setMinimumHeight(element.getMinHeight())
```

File: source/macos/AdaptiveCards/AdaptiveCards/Views/ACRContentStackView.swift [hunk 3]
```diff
@@ -5,7 +5,7 @@ protocol ACRContentHoldingViewProtocol {
     func addArrangedSubview(_ subview: NSView)
     func insertArrangedSubview(_ view: NSView, at insertionIndex: Int)
     func updateLayoutAndVisibilityOfRenderedView(_ renderedView: NSView, acoElement acoElem: ACSBaseCardElement, separator: SpacingView?, rootView: ACRView?)
-    func configureLayoutAndVisibility(_ verticalContentAlignment: ACSVerticalContentAlignment, minHeight: NSNumber?, heightType: ACSHeightType, type: ACSCardElementType, isRootMinHeightAvailable: Bool)
+    func configureLayoutAndVisibility(element: NSObject, isRootMinHeightAvailable: Bool)
```
Comment (urmitchauhan) [inline source/macos/AdaptiveCards/AdaptiveCards/Views/ACRContentStackView.swift:8]:
```suggestion
    func configureLayoutAndVisibility(element: ACSBaseCardElement, isRootMinHeightAvailable: Bool)
```

File: source/macos/AdaptiveCards/AdaptiveCards/Views/ACRContentStackView.swift [hunk 4]
```diff
@@ -316,26 +324,39 @@ class ACRContentStackView: NSView, ACRContentHoldingViewProtocol, SelectActionHa
     /// activation constraint all at once is more efficient than activating
     /// constraints one by one.
     
-    func configureLayoutAndVisibility(_ verticalContentAlignment: ACSVerticalContentAlignment, minHeight: NSNumber?, heightType: ACSHeightType, type: ACSCardElementType, isRootMinHeightAvailable: Bool) {
-        self.verticalContentAlignment = verticalContentAlignment
+    func configureLayoutAndVisibility(element: NSObject, isRootMinHeightAvailable: Bool) {
```
Comment (urmitchauhan) [inline source/macos/AdaptiveCards/AdaptiveCards/Views/ACRContentStackView.swift:327]:
```suggestion
    func configureLayoutAndVisibility(element: ACSBaseCardElement, isRootMinHeightAvailable: Bool) {
```

File: source/macos/AdaptiveCards/AdaptiveCards/Views/ACRContentStackView.swift [hunk 5]
```diff
@@ -545,6 +566,11 @@ extension ACRContentStackView: ACSVisibilityManagerFacade {
     func visibilityManagerSetLastStretchableView(isHidden: Bool) {
         visibilityManager.changeVisibilityOfLastStretchableView(isHidden: isHidden)
     }
+    
+    func visibilityManagerReactivateConstraint() {
```
Comment (urmitchauhan) [inline source/macos/AdaptiveCards/AdaptiveCards/Views/ACRContentStackView.swift:570]:
```suggestion
    func visibilityManagerUpdateConstraint() {
```

File: source/macos/AdaptiveCards/AdaptiveCards/Views/ACRView.swift [hunk 6]
```diff
@@ -179,12 +180,14 @@ class ACRView: ACRColumnView {
             }
             
             // toggle the last padding if all other
-            if facade?.visibilityManagerAllStretchableViewsHidden() ?? false {
-                facade?.visibilityManagerSetLastStretchableView(isHidden: !isHide)
-            } else {
-                facade?.visibilityManagerSetLastStretchableView(isHidden: isHide)
-            }
+            facade?.visibilityManagerSetLastStretchableView(isHidden: !(facade?.visibilityManagerAllStretchableViewsHidden() ?? false))
+            facadeArray.append(facade)
```
Comment (urmitchauhan) [inline source/macos/AdaptiveCards/AdaptiveCards/Views/ACRView.swift:184]: `visibilityManagerSetLastStretchableView(isHidden: isHide)` is depend on **isHide** variable.
can you please check?

File: source/macos/AdaptiveCards/AdaptiveCardsTests/Renderers/ColumnRendererTests.swift [hunk 7]
```diff
@@ -53,13 +53,14 @@ class ColumnRendererTests: XCTestCase {
         var columnView = renderColumnView()
         XCTAssertEqual(columnView.arrangedSubviews.count, 1)
         
+        // since we removed padding view all together and replaced with lastPadding but use paddingView for vertical content alignment this has changed
         column = .make(items: [FakeTextBlock.make()], verticalContentAlignment: .center)
         columnView = renderColumnView()
-        XCTAssertEqual(columnView.arrangedSubviews.count, 3)
+        XCTAssertEqual(columnView.arrangedSubviews.count, 4)
```
Comment (urmitchauhan) [inline source/macos/AdaptiveCards/AdaptiveCardsTests/Renderers/ColumnRendererTests.swift:59]: Why count is increase? if content at center then there is already two _stretchableViews_ are present, if i'm not wrong then why subviews count +1.

File: source/macos/AdaptiveCards/AdaptiveCardsTests/Renderers/ColumnRendererTests.swift [hunk 8]
```diff
@@ -53,13 +53,14 @@ class ColumnRendererTests: XCTestCase {
...
         
         column = .make(items: [FakeTextBlock.make()], verticalContentAlignment: .bottom)
         columnView = renderColumnView()
-        XCTAssertEqual(columnView.arrangedSubviews.count, 2)
+        XCTAssertEqual(columnView.arrangedSubviews.count, 3)
```
Comment (urmitchauhan) [inline source/macos/AdaptiveCards/AdaptiveCardsTests/Renderers/ColumnRendererTests.swift:63]: Why count is increase? if content at center then there is already one _**stretchableView**_ is present, if i'm not wrong then why subviews count +1.

File: source/macos/AdaptiveCards/AdaptiveCardsTests/Renderers/ContainerRendererTests.swift [hunk 9]
```diff
@@ -40,13 +41,14 @@ class ContainerRendererTests: XCTestCase {
         
         container = .make(verticalContentAlignment: .bottom, items: [FakeTextBlock.make()])
         containerView = renderContainerView(container)
+        // since we removed padding view all together and replaced with lastPadding but use paddingView for vertical content alignment this has changed
         // SpaceView 1
-        XCTAssertEqual(containerView.stackView.arrangedSubviews.capacity, 2)
+        XCTAssertEqual(containerView.stackView.arrangedSubviews.capacity, 3)
```
Comment (urmitchauhan) [inline source/macos/AdaptiveCards/AdaptiveCardsTests/Renderers/ContainerRendererTests.swift:46]: Please add comment for stretchable views is hide when there is other stretchable views are present.

File: source/macos/AdaptiveCards/AdaptiveCards/Views/ACRContentStackView.swift [hunk 10]
```diff
@@ -316,26 +324,39 @@ class ACRContentStackView: NSView, ACRContentHoldingViewProtocol, SelectActionHa
...
+                isBackgroundImageAvailable = true
+            }
+        } else if let collectionElement = element as? ACSCollectionTypeElement {
+            self.verticalContentAlignment = collectionElement.getVerticalContentAlignment()
+            minHeight = collectionElement.getMinHeight()
+            let bgImage = collectionElement.getBackgroundImage()
+            if let imageUrl = bgImage?.getUrl(), URL(string: imageUrl) != nil {
+                isBackgroundImageAvailable = true
+            }
+        }
```
Comment (urmitchauhan) [inline source/macos/AdaptiveCards/AdaptiveCards/Views/ACRContentStackView.swift:344]: can we move this `isBackgroundImageAvailable` at caller function. so we can remove the whole element dependency from this methods?
you can create tuple that can have necessary card properties.

```
typealias ACSElementProperties = (verticalContentAlignment: ACSVerticalContentAlignment, minHeight: NSNumber?, heightType: ACSHeightType, type: ACSCardElementType, isBackgroundImageAvailable: Bool)
```

```
func configureLayoutAndVisibility(elementProperties: ACSElementProperties, isRootMinHeightAvailable: Bool) {

}
```

**Review by urmitchauhan — CHANGES_REQUESTED** [review]

**Review by urmitchauhan — APPROVED** [review]

**Review by sidkotha — APPROVED** [review]

**Pull request merged**